split Client into Connection and Client

seanmonstar opened the pull request:
The Connection type is a `Future` that drives all of the IO of the
client connection.

The Client type is separate, and is used to send requests into the
connection.

An unresolved question is what order of the tuple is better:

- `let (client, conn) = await!(handshake)?;`
- `let (conn, client) = await!(handshake)?;`

carllerche commented on src/client.rs:
```diff
@@ -216,13 +206,40 @@ impl Default for Builder {
     T: AsyncRead + AsyncWrite,
 {
-    type Item = Client<T, B>;
+    type Item = (Connection<T, B>, Client<B>);
```
So, I don't really care about the order between `(Connection, Client)`, but I would like to try to be consistent.

I've been using `(Self, Other)` when adding a fn in a `impl Self`. 

That said, it seems more idiomatic in Rust to have fns that return tuples to be free fns (like `mpsc`). Though, I felt that was strange when the module was the same name as one of the types being returned...

seanmonstar commented:
> I believe that, as is, the PR would break Client::poll_ready behavior. Each Client handle is going to need its own Task to get notified when readiness transitions -> ready.

I haven't yet traced the `send_request` and `poll_ready` methods to the end, but I can see how this could be true. As a starting point, I added a test that send the `Client` to another thread, while waiting on the `Connection`. It seems to just work, but maybe that's a consequence of using the `mock` IO?

Also, re-ordered to `let (client, conn) = ...`.

seanmonstar commented on src/proto/streams/streams.rs:
```diff
@@ -442,6 +483,10 @@ where
 
         let stream = me.store.resolve(self.key);
+        if stream.is_pending_open {
+            //XXX: no?
```
This differs from the behavior when the user has an opened `Stream`, but tries to send more data than there is capacity in the window. We could stick with this (with a different error variant), or allow buffering data even on a pending stream, to be more consistent for the user.

seanmonstar commented on src/proto/streams/streams.rs:
```diff
@@ -625,7 +680,15 @@ where
+        let mut me = match self.inner.lock() {
+            Ok(inner) => inner,
+            Err(_) => if ::std::thread::panicking() {
+                trace!("StreamRef::drop; mutex poisoned");
```
I noticed quite a few times when hitting assertions that this drop function was causing aborts. This code could likely stay, because aborting is 💩 .

seanmonstar commented on tests/support/mod.rs:
```diff
@@ -1,5 +1,11 @@
 //! Utilities to support tests.
 
+#[cfg(not(feature = "unstable"))]
+compile_error!(
```
If you forget to enable the `unstable` feature when running the tests, the error message is quite bizarre (about duplicate functions on `frames::Mock` and so forth). We can make it super clear with this.

seanmonstar commented on src/client.rs:
```diff
@@ -91,7 +100,7 @@ where
     /// stream.
     pub fn poll_ready(&mut self) -> Poll<(), ::Error> {
-        self.connection.poll_send_request_ready()
+        self.inner.poll_pending_open(self.pending.as_ref())
```
We may want to add a `fn ready(self) -> impl Future` to ease dealing with this.

I also think it may be nice to add a similar future to `Stream` for when capacity is available (maybe combines with when the pending stream is opened, and so a user could send data, depends on dicussion lower in this PR).

seanmonstar commented on src/proto/streams/prioritize.rs:
```diff
@@ -476,122 +484,142 @@ where
+                        if let Err(_e) = stream.state.send_open(end_stream) {
+                            // other peer may have closed this stream between us pending
+                            // and finally opening
+                            error!("pending open errored");
```
This may not be great. I originally had `send_open(..).expect("this shouldn't happen")`...

seanmonstar commented:
Updated with in-person comments.

- Opening a pending stream is almost just like opening a new stream. If there is no room for another concurrent stream, it is added to a `pending_open` queue.
- Moved popping of the queue out of `pop_frame`, and into the loop of `poll_complete`.

carllerche commented:
Oh rustfmt
